### ISSUE TITLE
[BUG] Fields are not correctly inherited from parent to child in dataclasses

### ISSUE THREAD

**Issue by Yurzs** [cython/cython#4799]
**Describe the bug**
There is no way to inherit default values correctly in subclasses of a dataclass when using `.pxd`. There are two cases with not clear behaviour:
* Case1. If you place `@cython.dataclasses.dataclass` on cdef classes in `.pxd` then fields from parent are added to child but default values are lost and you are forced to fill this value.
* Case2. You place `@cython.dataclasses.dataclass` only in `.pyx` then field is missing from `__init__` and is present in object as `None`

**To Reproduce**
Code to reproduce the behaviour:

`common.document.pxd`:
```cython
cdef class Document:
    cdef public str _id
```

`common.document.pyx`:
```cython
import cython

@cython.dataclasses.dataclass(kw_only=True)
@cython.cclass
cdef class Document:
    _id: str = cython.dataclasses.field(default_factory=lambda: str(bson.ObjectId()))
```

`common.models.account.pxd`:
```cython
import cython

from common.document cimport Document

@cython.cclass
cdef class Account(Document):
    cdef public int id
    cdef public str username
    cdef public str password

    cdef public char status
```

`common.models.account.pyx`:
```cython
import cython

from common.document import Document

@cython.dataclasses.dataclass(kw_only=True)
@cython.cclass
cdef class Account(Document):
    id: cython.int
    username: str
    password: str
    
    status: cython.char = 0
```

```python
import pyximport
pyximport.install(language_level=3, build_in_temp=True)

 from common.model.account import Account
 
 acc = Account(id=1, username="test", password="password")
 print(acc._id) # None
```

You can check Case1 by adding to `@cython.dataclasses.dataclass(kw_only=True)` to `Account` in `account.pxd`

**Expected behavior**
Fields with all their properties should be present in child as in parent.

**Environment (please complete the following information):**
 - OS: [macOS 12.3]
 - Python version [3.10.0]
 - Cython version [3.0.0a10]

**Comment by da-woods** [comment]
Inheritance (within the same pyx file) is tested for and I believe works: https://github.com/cython/cython/blob/d9709b249f4a1dd73946f5077866a1b81ba0b128/tests/run/cdef_class_dataclass.pyx#L80

I think I completely forgot about pxd files when I originally wrote the dataclasses feature.

I suspect the information does need to go into the pxd file (i.e. put the dataclass decorator in class in the pxd file) but it needs a bit of thought about the best way to implement that. I think it'd need the full field definition in the pxd file in order to generate the constructor correctly (since it'd have to know about `default_factory` at compile-time). That doesn't really fit with how Cython generally works though...

**Comment by da-woods** [comment]
Some of it could probably be solved by just calling the base-class methods (and thus avoiding needing to know about the inheritance) but getting the function signature for the constructor right seems like it might be hard...

I'm currently tempted to just make it an error to inherit a cdef dataclass from a cdef dataclass in a different file until we have a good solution. That isn't a good solution, but it's better than behaving unexpectedly.

**Comment by Yurzs** [comment]
Yeah, everything works if every dataclass is placed in a single file.
Compile time goes through the roof though. 
Its taking several minutes for a 1k line file (with 30 dataclasses) on m1 pro mac

**Comment by da-woods** [comment]
We now at least warn about this. I'd quite like a better solution, but I think inheriting cdef dataclasses across files is going to be hard to implement (because most of the information is in the pyx file rather than the pxd file)
